Tests/NegativeCompileTestsRunner/NegativeTestRunner.swift

```diff
@@ -0,0 +1,389 @@
+        #if !DEBUG
+        arguments += " -c release -Xswiftc -enable-testing"
+        #endif
+        #if os(Linux)
+        arguments += " --enable-test-discovery"
+        #endif
+        #if COVERAGE // custom defined Active Compilation Condition which we set when we enable code coverage collection
+        arguments += " --enable-code-coverage -Xswiftc -DCOVERAGE"
+        #endif
+
+        let stdOutput = try runCommand(command: "swift", arguments: arguments, expectedStatus: 1)
+
+        print("[\(identifier)] Scanning results for target \(target.name)...")
+
+        let lines = stdOutput.split(separator: "\n", omittingEmptySubsequences: false).map {
+            String($0)
+        }
+
+        for line in lines {
+            let matches = compilerErrorPattern.matches(in: line, range: NSRange(line.startIndex..., in: line))
+
+            guard let match = matches.first, matches.count == 1 else {
+                continue
+            }
+
+            let filePath = line.retrieveMatch(match: match, at: 1)
+            let lineNumber = Int(line.retrieveMatch(match: match, at: 2)) ?? -1
+            let column = Int(line.retrieveMatch(match: match, at: 4))
+            let errorMessage = line.retrieveMatch(match: match, at: 5)
+
+            let marked = markExpectedErrorTriggered(in: target, absolutePath: filePath, line: lineNumber, error: errorMessage)
+            if !marked {
+                self.unexpectedErrors.insert(
+                    UnexpectedError(filePath: filePath, line: lineNumber, column: column, errorMessage: errorMessage, rawLine: line)
+                )
+            }
+        }
+    }
+
+    @discardableResult
+    private func runCommand(command: String, arguments: String, expectedStatus: Int32 = 0) throws -> String {
+        var parts: [String] = []
+
+        print("-----------------------------")
+        print("Running command '\(command) \(arguments)'...")
+
+        guard let swiftBinary = Task.findExecutable(named: command) else {
+            fatalError("Could not find '\(command)' executable!")
+        }
+
+        // https://stackoverflow.com/questions/67595371/swift-package-calling-usr-bin-swift-errors-with-failed-to-open-macho-file-to
+        var environment: [String: String] = ProcessInfo.processInfo.environment
+        if environment.keys.contains("OS_ACTIVITY_DT_MODE") {
+            print("Clearing OS_ACTIVITY_DT_MODE environment variable")
+            environment["OS_ACTIVITY_DT_MODE"] = nil
+        }
+
+        let task = Task(
+            executableUrl: swiftBinary,
+            arguments: arguments.split(separator: " ").map { String($0) },
+            workingDirectory: workingDirectory,
+            captureOutput: true,
+            redirectStderrToStdout: true,
+            launchInCurrentProcessGroup: false,
+            environment: environment,
+            inheritsParentEnvironment: false // even though we pass false, our construction made above inherits environment from parent!
```

Comment:
The reason for this is that the `environment` you're passing to the initialiser was initially copied from `ProcessInfo.processInfo.environment`.

The `Task` API, if `inheritsParentEnvironment` is false, will set the child's environment to what was passed to the initialiser (or set after initialising but before launching the process).

Or am I missing something here?

Reply:
Thats exactly the intention here.
We want to inherit the environment form the parent process, but need to unset a certain environment variable in certain conditions, see here https://github.com/Apodini/Apodini/blob/96d12e977afd0f1105d135a10976ee2b8cff1734/Tests/NegativeCompileTestsRunner/NegativeTestRunner.swift#L320-L325 (this is to workaround an issue when running the TestRunner inside Xcode).

To me it seemed that with `inheritsParentEnvironment` the Task API only supports adding additional environment variables on top, but doesn't have an option to **unset** certain variables, or am I wrong?

Reply:
The way it's implemented is that, based on the value of `inheritsParentEnvironment`, the child will get either only the environment passed to the initialiser (if it shouldn't inherit), or the current environment merged with what was passed to the task (if it should inherit).

https://github.com/Apodini/Apodini/blob/7b26ea0397b5e66eee2fc7cf035636a7e778e5fe/Sources/ApodiniUtils/Task.swift#L179-L186

It's correct that if `inheritsParentEnvironment` is true, you wouldn't be able to unset something from the parent (due to how the merging works, given that the `environment` dictionary's `Value` is `String` instead of `String?`.
But if `inheritsParentEnvironment` is false, it simply sets the task's environment dictionary as the child's environment, which (at least when I tested it locally when I wrote it) should completely ignore the parent's Process.environment.

Feel free to change this though (I think simply changing `Task.environment`'s type to `[String: String?]` and thus using nil values to make sure the for loop deletes existing values should suffice??)

If you want to see the actual environment being passed to the child you can have a look at `CApodiniUtils.c`, that's where the pseudo-child is replaced with the proper child (note that this only applies to Tasks where `launchInCurrentProcessGroup == true`)

https://github.com/Apodini/Apodini/blob/7b26ea0397b5e66eee2fc7cf035636a7e778e5fe/Sources/CApodiniUtils/CApodiniUtils.c#L45

Reply:
Are you running this in Xcode or via the command line?
I think that might also make a difference since Xcode/lldb will in some cases interfere with the process structure

Reply:
> The way it's implemented (or, was intended to be implemented) is that, based on the value of inheritsParentEnvironment, the child will get either only the environment passed to the initialiser (if it shouldn't inherit), or the current environment merged with what was passed to the task (if it should inherit).

Yes I noticed that. 

> It's correct that if inheritsParentEnvironment is true, you wouldn't be able to unset something from the parent (due to how the merging works, given that the environment dictionary's Value is String instead of String?.
But if inheritsParentEnvironment is false, it simply sets the task's environment dictionary as the child's environment, which (at least when I tested it locally when I wrote it) should completely ignore the parent's Process.environment.

> Feel free to change this though (I think simply changing Task.environment's type to [String: String?] and thus using nil values to make sure the for loop deletes existing values should suffice??)

You are right, changing that to a Dictionary with optional Strings could probably be used to encode "unset variable" using `nil`. Haven't really though about changing the `Task` internals. As this is probably a pretty unique use case and also just used for another workaround I'm not sure if we really need to adapt the `Task` API and instead leave it as is(?).

> Are you running this in Xcode or via the command line?

Well, yes 🙃 Both should work. In the CI (or if you wish to run it locally) it should be executable via command line. But tests are also often executed within Xcode (or any other IDE), therefore running inside Xcode is also supported.
Both work right now.

Reply:
No, `Task` should absolutely be changed to include this functionality (after all, what's the point of having a child process API with support for setting the child's environment, if it doesn't let you set the child's environment?).

It is, however, still weird that this is happening, since you're setting `inheritsParentEnvironment` to false, meaning it never enters the first branch of the if stmt (the one w/ the loop) and instead just sets the environment to the dictionary passed in (which does not contain the key bc it was deleted by the caller)
Maybe swiftc modifies its own environment to re-set the key?

I can have a look at this if you want.
You could also try changing it to launch in the current process group, update the C file linked above to dump its envp, and run the test in the terminal. That'd give you the definitive environment passed to the child, since you'd be reading the child's envp as actually passed by Foundation's Process (i.e. after anyone could change it)

Reply:
> It is, however, still weird that this is happening [...]

Not sure what you are talking about? Everything works as expected.

Reply:
The way I understood this the issue was that, despite passing in a custom environment with a specific key-value pair deleted, the child still ended up being launched with that key-value pair present?
At least that's how I interpreted this comment:
https://github.com/Apodini/Apodini/blob/96d12e977afd0f1105d135a10976ee2b8cff1734/Tests/NegativeCompileTestsRunner/NegativeTestRunner.swift#L335

Reply:
Nope.
This is just a disclaimer/heads up to the reader who shouldn't be fooled by the `inheritsParentEnvironment: false`, as we **do** inherit from the parent environment, because we construct that ourselves above (we just do not really on the environment construction done by `Task`).
And the reason we construct the parent environment ourselves was because of the introduce workaround to delete a specific key form the environment in certain situations.

Reply:
Ooh ok well in that case you probably can ignore everything I said above (excl the making it `[String: String?]` thing)

Reply:
Solved via c9be844